File: 0022-Generate-Parentheses/java-0022/src/Solution1.java
/// Source : https://leetcode.com/problems/generate-parentheses/
/// Author : penpenps
/// Time   : 2019-05-28

import java.util.*;

/// Recursive with "(" and ")" separately and guarantee open and close num is equal
/// Time Complexity: O(2^n)
/// Space Complexity: O(n)

public class Solution1 {
    public void helper(List<String> ans, String str, int open, int close, int max) {
        if(str.length() == 2*max){
            ans.add(str);
            return;
        }

        if(open < max)
            helper(ans, str+"(", open+1, close, max);
        if(close < open)
            helper(ans, str+")", open, close+1, max);
    }
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(ans, "", 0, 0, n);
        return ans;
    }

    private static void printList(List<String> list){
        for(String e: list)
            System.out.print(e + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        int n = 3;
        List<String> res = (new Solution1()).generateParenthesis(n);
        printList(res);
    }
}

File: 0022-Generate-Parentheses/py-0022/Solution1.py
# Source : https://leetcode.com/problems/generate-parentheses/
# Author : penpenps
# Time   : 2019-05-28

from typing import List

# Recursive with "(" and ")" separately and guarantee open and close num is equal
# Time Complexity: O(2^n)
# Space Complexity: O(n)

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def helper(current, openNum, closeNum, n, ans):
            if len(current) == 2*n:
                ans.append(current)
                return
            if openNum < n:
                helper(current+"(", openNum + 1, closeNum, n, ans)
            if closeNum < openNum:
                helper(current+")", openNum, closeNum + 1, n, ans)
                
        ans = []
        helper("", 0, 0, n, ans)
        return ans

if __name__ == '__main__':
	solution = Solution()
	n = 3
	print(solution.generateParenthesis(n))